apollo-router/tests/apollo_reports.rs

```diff
@@ -15,74 +17,105 @@ use flate2::read::GzDecoder;
 use http::header::ACCEPT;
 use once_cell::sync::Lazy;
 use prost::Message;
+use prost_types::Timestamp;
+use proto::reports::trace::Node;
+use serde_json::json;
 use tokio::sync::Mutex;
+use tokio::sync::OnceCell;
 use tower::Service;
 use tower::ServiceExt;
 use tower_http::decompression::DecompressionLayer;
 
+use crate::proto::reports::trace::node::Id::Index;
+use crate::proto::reports::trace::node::Id::ResponseName;
 use crate::proto::reports::Report;
+use crate::proto::reports::Trace;
 
 static REPORTS: Lazy<Arc<Mutex<Vec<Report>>>> = Lazy::new(Default::default);
 static TEST: Lazy<Arc<Mutex<bool>>> = Lazy::new(Default::default);
 static ROUTER_SERVICE_RUNTIME: Lazy<Arc<tokio::runtime::Runtime>> = Lazy::new(|| {
     Arc::new(tokio::runtime::Runtime::new().expect("must be able to create tokio runtime"))
 });
+static CONFIG: OnceCell<serde_json::Value> = OnceCell::const_new();
 static ROUTER_SERVICE: Lazy<Arc<Mutex<Option<BoxCloneService>>>> = Lazy::new(Default::default);
+static MOCKED_ROUTER_SERVICE: Lazy<Arc<Mutex<Option<BoxCloneService>>>> =
+    Lazy::new(Default::default);
 
-async fn get_router_service() -> BoxCloneService {
-    let mut router_service = ROUTER_SERVICE.lock().await;
-    if router_service.is_none() {
-        let reports = &*REPORTS;
-        std::env::set_var("APOLLO_KEY", "test");
-        std::env::set_var("APOLLO_GRAPH_REF", "test");
-
-        let listener = std::net::TcpListener::bind("127.0.0.1:0").unwrap();
-        let addr = listener.local_addr().unwrap();
-        let app = axum::Router::new()
-            .route("/", post(report))
-            .layer(DecompressionLayer::new())
-            .layer(tower_http::add_extension::AddExtensionLayer::new(
-                reports.clone(),
-            ));
-
-        let mut config: serde_json::Value =
-            serde_yaml::from_str(include_str!("fixtures/apollo_reports.router.yaml"))
-                .expect("apollo_reports.router.yaml was invalid");
-        config = jsonpath_lib::replace_with(config, "$.telemetry.apollo.endpoint", &mut |_| {
-            Some(serde_json::Value::String(format!("http://{addr}")))
+async fn config() -> serde_json::Value {
+    CONFIG
+        .get_or_init(|| async {
+            std::env::set_var("APOLLO_KEY", "test");
```

Comment:
I'm not sure about setting these env variables. Some other tests may rely on these values being valid.